src/main/scala/com/lohika/vteraz/validator/Validator.scala

```diff
@@ -0,0 +1,118 @@
+package com.lohika.vteraz.validator
+
+/**
+  * Implement validator typeclass that should validate arbitrary value [T].
+  *
+  * @tparam T the type of the value to be validated.
+  */
+trait Validator[T] {
+    /**
+      * Validates the value.
+      *
+      * @param value value to be validated.
+      * @return Right(value) in case the value is valid, Left(message) on invalid value
+      */
+    def validate(value: T): Either[String, T]
+
+    /**
+      * And combinator.
+      *
+      * @param other validator to be combined with 'and' with this validator.
+      * @return the Right(value) only in case this validator and <code>other</code> validator returns valid value,
+      *         otherwise Left with error messages from the validator that failed.
+      */
+    def and(other: Validator[T]): Validator[T] = {
+        value: T => {
+            val firstResult = validate(value)
+            if (firstResult.isLeft) {
+                Left(firstResult.left.get)
+            } else {
+                val secondResult = other.validate(value)
+                if (secondResult.isLeft) {
+                    Left(secondResult.left.get)
+                } else {
+                    Right(value)
+                }
+            }
+        }
+    }
+
+    /**
+      * Or combinator.
+      *
+      * @param other validator to be combined with 'or' with this validator.
+      * @return the Right(value) only in case either this validator or <code>other</code> validator returns valid value,
+      *         otherwise Left with error messages from both validators.
+      */
+    def or(other: Validator[T]): Validator[T] = {
+        value: T => {
+            val firstResult = validate(value)
+            val secondResult = other.validate(value)
+            val groupedResults = List(firstResult, secondResult).groupBy(result => result.isRight)
```

Comment:
Was creating a list of two values really necessary? You could use `pattern matching` to simplify your code (or use separate checks)